### PR TITLE
PBR -Itamar

### LINKED ISSUE #179: PBR shaders
i as a dev want to write PBR shaders to make the graphics look better

### LINKED ISSUE #221: Update light struct to shader
The editor controls to alter light color exists today, but is not hooked up to functionality. Update the light struct to pass the data for use in shaders.

### LINKED ISSUE #223: Move gamma correction into PBR shaders

### LINKED ISSUE #224: Gamma correction in PBR
As a technical artist I want to include gamma correction in PBR to increase performance by not using a separate pass.

### PR DESCRIPTION
**Pull request by Ittis**
Implementerat en PBR shader + fixat olika imGUI grejer relaterade till PBR och även fixat så det går att sätta färg på ljuset. 

Resolves #179 
Fixes #224 
Fixes #223 
Fixes #221

### CODE REVIEW

**Comment by NinjaDanz3r** [comment]
Can one of the admins verify this patch?

File: shaders/DefaultPBR.frag [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+} vertexIn;
+
+//uniform sampler2D baseImage;
+//uniform sampler2D normalMap;
+//uniform sampler2D specularMap;
+//uniform sampler2D glowMap;
+
+//uniform vec3 albedo;
+//uniform float metallic;
+//uniform float roughess;
```
Comment (ludvigarlebrink) [inline shaders/DefaultPBR.frag]: Remove out commented code?

File: shaders/DefaultPBR.frag [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+in VertexData {
+    vec3 normal;
+    vec3 tangent;
+    vec2 texCoords;
+} vertexIn;
+
+//uniform sampler2D baseImage;
+//uniform sampler2D normalMap;
+//uniform sampler2D specularMap;
+//uniform sampler2D glowMap;
```
Comment (ludvigarlebrink) [inline shaders/DefaultPBR.frag]: Remove out commented code?

File: shaders/DefaultPBR.frag [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+//uniform float metallic;
+//uniform float roughess;
+
+vec3 albedo = (0.5f,0.5f,0.5f);
+float metallic = 0.2f;
+float roughess = 0.5f;
+
+
+
+
```
Comment (ludvigarlebrink) [inline shaders/DefaultPBR.frag]: Remove some newlines?

File: shaders/Deferred.frag [hunk 4]
```diff
@@ -31,67 +33,135 @@ uniform Light lights[maxLights];
...
+		Lo  += (kD * albedo / PI + specular) * radiance * NdotL;
+		Lo += lights[i].ambientCoefficient * albedo;
+	}
+
+	return Lo;
+}
+
+
+void main()
+{
```
Comment (ludvigarlebrink) [inline shaders/Deferred.frag]: Keep void main()  { on same line.

File: shaders/Deferred.frag [hunk 5]
```diff
@@ -31,67 +33,135 @@ uniform Light lights[maxLights];
...
+		vec3 kS = F;
+		vec3 kD = vec3(1.0f) - kS;
+		kD *= 1.0f - metallic;
+		
+		float NdotL = max(dot(N,surfaceToLight), 0.0f);
+		Lo  += (kD * albedo / PI + specular) * radiance * NdotL;
+		Lo += lights[i].ambientCoefficient * albedo;
+	}
+
+	return Lo;
```
Comment (ludvigarlebrink) [inline shaders/Deferred.frag]: Remove tabs keep spaces.

File: src/Video/Lighting/Lighting.cpp [hunk 6]
```diff
@@ -117,10 +117,14 @@ void Lighting::Render(const glm::mat4& inverseProjectionMatrix) {
...
     // Set uniforms.
-    glUniform1i(shaderProgram->GetUniformLocation("tDiffuse"), Lighting::DIFFUSE);
-    glUniform1i(shaderProgram->GetUniformLocation("tNormals"), Lighting::NORMAL);
-    glUniform1i(shaderProgram->GetUniformLocation("tSpecular"), Lighting::SPECULAR);
-    glUniform1i(shaderProgram->GetUniformLocation("tGlow"), Lighting::GLOW);
+    //glUniform1i(shaderProgram->GetUniformLocation("tDiffuse"), Lighting::DIFFUSE);
+    //glUniform1i(shaderProgram->GetUniformLocation("tNormals"), Lighting::NORMAL);
+    //glUniform1i(shaderProgram->GetUniformLocation("tSpecular"), Lighting::SPECULAR);
+    //glUniform1i(shaderProgram->GetUniformLocation("tGlow"), Lighting::GLOW);
+	glUniform1i(shaderProgram->GetUniformLocation("textureAlbedo"), Lighting::DIFFUSE);
```
Comment (ludvigarlebrink) [inline src/Video/Lighting/Lighting.cpp]: Remove tabs keep spaces.

File: src/Video/Lighting/Lighting.hpp [hunk 7]
```diff
@@ -72,6 +72,7 @@ namespace Video {
                 GLint ambientCoefficient;
                 GLint coneAngle;
                 GLint direction;
+				GLint color;
```
Comment (ludvigarlebrink) [inline src/Video/Lighting/Lighting.hpp]: Remove tabs keep spaces.

File: shaders/Deferred.frag [hunk 8]
```diff
@@ -31,67 +33,135 @@ uniform Light lights[maxLights];
...
+	float NdotV = max(dot(N,V), 0.0);
+	float NdotL = max(dot(N,L), 0.0);
+	float ggx1 = GemetrySchlickGGX(NdotV, roughness);
+	float ggx2 = GemetrySchlickGGX(NdotL, roughness);
+
+	return clamp((ggx1 * ggx2),0,1);
+}
+
+vec3 ReconstructPos(in vec2 texCoord, in float depth)
+{
```
Comment (ludvigarlebrink) [inline shaders/Deferred.frag]: vec3 ReconstructPos(in vec2 texCoord, in float depth) {

File: shaders/Deferred.frag [hunk 9]
```diff
@@ -31,67 +33,135 @@ uniform Light lights[maxLights];
...
-    gl_FragDepth = depth;
+	float depth = texture(tDepth, vertexIn.texCoords).r;
+	vec3 test = texture(textureAlbedo, vertexIn.texCoords).rgb;
+	vec3 albedo	= pow(test, vec3(2));
+	vec3 normal	= normalize(texture(textureNormal, vertexIn.texCoords).rgb);
+	float metallic = texture(textureMetallic, vertexIn.texCoords).r;
+	float roughness = texture(textureRougness, vertexIn.texCoords).r;
+	vec3 pos = ReconstructPos(vertexIn.texCoords, depth);
+
+vec3 applyLights()
```
Comment (ludvigarlebrink) [inline shaders/Deferred.frag]: vec3 applyLights() Capital case on functions (vec3 ApplyLights())

File: shaders/Deferred.frag [hunk 10]
```diff
@@ -31,67 +33,135 @@ uniform Light lights[maxLights];
...
-        specularCoefficient = pow(max(dot(surfaceToCamera, reflect(-surfaceToLight, normal)), 0.0), materialShininess);
-    vec3 specular = specularCoefficient * surfaceSpecular * lights[light].intensities;
-    
-    // Linear color (color before gamma correction)
-    return ambient + attenuation * (diffuse + specular);
+const float PI = 3.14159265359;
+
+vec3 FresnelSchlick(float cosTheta, vec3 F0)
+{
+	return F0 + (1.0 - F0) * pow(1.0 - cosTheta, 5.0);
```
Comment (LarsIngo) [inline shaders/Deferred.frag]: Don't use pow, it's really slow

File: shaders/DefaultPBR.frag [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+layout(location = 0) out vec3 diffuseOut;
+layout(location = 1) out vec3 normalsOut;
+layout(location = 2) out vec3 specularOut;
+layout(location = 3) out vec3 glowOut;
+
+// Calculate normal based on interpolated vertex normal, sampled normal (from normal map) and vertex tangent.
+vec3 calculateNormal(in vec3 normal, in vec3 tangent, in vec3 mapNormal) {
```
Comment (ludvigarlebrink) [inline shaders/DefaultPBR.frag]: Capital case on functions.

File: shaders/Deferred.frag [hunk 12]
```diff
@@ -31,67 +33,135 @@ uniform Light lights[maxLights];
...
-    for (int i=0; i < lightCount; ++i) {
-        accumulatedLight += ApplyLight(diffuse, normal, position, specular, i);
-    }
-    
-    fragmentColor = vec4(accumulatedLight, 1.0);
-    extraOut = vec4(texture(tGlow, vertexIn.texCoords).rgb, 1.0);
-    gl_FragDepth = depth;
+	float depth = texture(tDepth, vertexIn.texCoords).r;
+	vec3 test = texture(textureAlbedo, vertexIn.texCoords).rgb;
+	vec3 albedo	= pow(test, vec3(2));
```
Comment (LarsIngo) [inline shaders/Deferred.frag]: "test". Old name? Try not using pow

File: shaders/Deferred.frag [hunk 13]
```diff
@@ -31,67 +33,135 @@ uniform Light lights[maxLights];
...
-    if (diffuseCoefficient > 0.0)
-        specularCoefficient = pow(max(dot(surfaceToCamera, reflect(-surfaceToLight, normal)), 0.0), materialShininess);
-    vec3 specular = specularCoefficient * surfaceSpecular * lights[light].intensities;
-    
-    // Linear color (color before gamma correction)
-    return ambient + attenuation * (diffuse + specular);
+const float PI = 3.14159265359;
+
+vec3 FresnelSchlick(float cosTheta, vec3 F0)
+{
```
Comment (LarsIngo) [inline shaders/Deferred.frag]: { should be on the same line as function

File: shaders/Deferred.frag [hunk 14]
```diff
@@ -31,67 +33,135 @@ uniform Light lights[maxLights];
...
+	return F0 + (1.0 - F0) * pow(1.0 - cosTheta, 5.0);
+}
+
+float DistributionGGX(vec3 N, vec3 H, float roughness)
+{
```
Comment (LarsIngo) [inline shaders/Deferred.frag]: { should be on the same line as function

File: shaders/Deferred.frag [hunk 15]
```diff
@@ -31,67 +33,135 @@ uniform Light lights[maxLights];
...
+	float nom = a2;
+	float denom	=(NdotH2 * (a2 - 1.0f) + 1.0f);
+	denom = PI * denom * denom;
+	
+	//return nom/denom;
+	return clamp((nom /denom),0,5.5);
+}
+
+float GemetrySchlickGGX(float NdotV, float roughness)
+{
```
Comment (LarsIngo) [inline shaders/Deferred.frag]: { should be on the same line as function

File: shaders/Deferred.frag [hunk 16]
```diff
@@ -31,67 +33,136 @@ uniform Light lights[maxLights];
...
+    return nom/denom;
 }
 
-// Reconstruct position.
-vec3 ReconstructPos(in vec2 texCoord, in float depth){
-    vec4 sPos = vec4(texCoord * 2.0 - 1.0, depth * 2.0 - 1.0, 1.0);
+float GeometrySmith(vec3 N, vec3 V, vec3 L, float roughness) {
+    float NdotV = max(dot(N,V), 0.0f);
+    float NdotL = max(dot(N,L), 0.0f);
+    float ggx1 = GemetrySchlickGGX(NdotV, roughness);
```
Comment (Sajdkick) [inline shaders/Deferred.frag]: GemetrySchlickGGX -> Ge_**o**_metrySchlickGGX?

**Pull request merged**